### PR TITLE
Fix token corruption causing major client slowdown

### PR DESCRIPTION
**Pull request by stevenh**
Due to token corruption in the server cache all server connections where being passed an empty token, causing major client slowdown presumably while the server checked the invalid token.

Correct length of string generated by CPlexAES::encrypt, which was causing token corruption issue.

Use AES_BLOCK_SIZE instead of literal 16

Optimise empty input case for both CPlexAES encrypt and decrupt.

Ensure we pass a valid AES_BLOCK_SIZE size block to CPlexAES encrypt and decrupt to ensure we don't have buffer overruns.

Close file to prevent leak in decryptFile.

Use IsEmpty() instead of empty() since .. CStdString.

Optimise token loop in BuildURL.

### CODE REVIEW

**Comment by stevenh** [comment]
As users cache files contain corrupted token information due to this bug, they will need to delete the cache file after updating to fully fix the problem.

On windows 7 this is located here:
%APPDATA%\Plex Home Theater\userdata\Database\PlexServerCache3.db

**Comment by tru** [comment]
Hello @stevenh! Great! Thanks for tracking down this issue!

Can I suggest that you split the changes into two? One that fixes the AES_BLOCK_SIZE and the other one that makes sure we don't append a empty token? This way it's way easier to see what has changed and we don't hide important fixes between a lot of things.

Also a lot of removed empty new lines seems to have snuck into this, was that intentional?

**Comment by stevenh** [comment]
The empty line changes aren't removals, they are actually just removing trailing white space, its something that my vim setup does automatically as all the other projects I work on frown on it; hope this is OK?

**Comment by tru** [comment]
Yeah that sounds fine. I am just allergic to style fixes mixed with real changes since it looks ugly and can be hard to follow when you audit your history, but I'll let it slip this time ;)

**Comment by stevenh** [comment]
Split it into two commits as requested, hope this helps :)

**Comment by tru** [comment]
Perfect! Thanks a lot. I will merge tomorrow.

**Comment by tru** [comment]
Merged to internal tree. Will be available in 1.3.6, thanks again!

**Comment by tru** [comment]
@stevenh just wanted to let you know now that I have pushed 1.3.6 to the public repo that I had to back out some of the changes you made since they lead to other corruption. I discovered it really really close to the release so it was way safer for me to back out the changes to AES decryption than to figure out what was really wrong. I have added a test to test encrypt/decrypt now as well. If you want to have a second look I am happy to see if we can find something that works everywhere. Just look at my commit history for PlexAES.cpp and you'll see what I had to do.

**Comment by stevenh** [comment]
Thanks for the heads up @tru and for adding the tests.

I've revisited it and can see what the issue was, storing the shortened final block was the issue as decrypt requires the entire block to decode.

The fix needs to be done via changes in decrypt instead.

I've fixed this, updating the tests to confirm as well in my new pull https://github.com/plexinc/plex-home-theater-public/pull/25
